Fix Parquet with special characters in field names.

rdblue opened the pull request:
This uses Avro's name sanitization methods to ensure that field names are compatible with Parquet. The names stored in each file don't actually matter because Iceberg uses field IDs.

rdsr commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -347,6 +347,13 @@ static Schema copyRecord(Schema record, List<Schema.Field> newFields, String new
     return copy;
   }
 
+  public static String makeCompatibleName(String name) {
```
Nit: do you want to use this method in org.apache.iceberg.avro.TypeToSchema#struct() api? 
Currently the code is 
```
boolean isValidFieldName = AvroSchemaUtil.validAvroName(origFieldName);
String fieldName =  isValidFieldName ? origFieldName : AvroSchemaUtil.sanitize(origFieldName);
```

rdblue replied:
I considered it, but the `isValidFieldName` is reused to add the original name as a property, so I think it's fine as it is.

rdblue commented:
@rdsr, please have another look. I added tests to iceberg-data and ended up needing to fix a couple of things:

* `BuildAvroProjection` couldn't project fields with special characters in the name because Avro would reject the projection schema. I added a couple calls to `makeCompatibleName` to fix this. It should be safe because special characters would cause failures in these cases before.
* The generic `DataReader` created iceberg-data records with a schema converted from Avro. When using sanitized names, the records would not use the right field names for `getField`. The fix was to pass in the original Iceberg schema. To make this easy, I added a new visitor.

I think the second fix also addresses the case introduced by #207 when the Avro names don't match because the shouldn't be projected. Next, we should be able to fix Avro reads by using a similar pattern to iceberg-data, but one that produces Avro generics.

rdsr commented:
Thanks @rdblue . I'll have a look over the weekend

rdsr commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  private static <T> T visitArray(Type type, Schema array, AvroSchemaWithTypeVisitor<T> visitor) {
+    if (array.getLogicalType() instanceof LogicalMap) {
```
Is it worth while to also check for `AvroSchemaUtil.isKeyValueSchema(array.getElementType())`?

rdblue replied:
No, but I think it would be good to check whether the Iceberg type is a map.

rdsr commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java:
```diff
@@ -0,0 +1,132 @@
+  public static <T> T visit(Type iType, Schema schema, AvroSchemaWithTypeVisitor<T> visitor) {
+    switch (schema.getType()) {
+      case RECORD:
+        return visitRecord(iType != null ? iType.asStructType() : null, schema, visitor);
```
what could the expected type be null? I see that we are traversing `NULL` branch of the union, is it because of that?

rdsr commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java:
```diff
@@ -0,0 +1,132 @@
+    return visit(iSchema.asStruct(), schema, visitor);
+  }
+
+  public static <T> T visit(Type iType, Schema schema, AvroSchemaWithTypeVisitor<T> visitor) {
```
When could the expected type be null? I see that we are traversing `NULL` branch of the union, is it because of that? Also, I'm not clear on why do we need to visit the `NULL` branch of the union

rdblue replied:
The Iceberg type might be null if the Avro type has no corresponding field. For example, if we drop a column from an Iceberg schema and read an older data file, that column will not be in the read schema, but will be in file schemas.

rdblue commented:
Merged. Thanks for reviewing, @Parth-Brahmbhatt and @rdsr!
